src/memory/redis.rs

```diff
@@ -0,0 +1,253 @@
+    index_name: String,
+}
+
+#[async_trait]
+impl MemorySystem for RedisMemorySystem {
+    async fn store_memory(&mut self, llm: &LLM, memory: &str) -> Result<(), Box<dyn Error>> {
+        let embedding = llm.model.get_base_embed(memory).await?;
+        let mut con = self.client.get_tokio_connection().await?;
+
+        let memory_struct = RedisPayload {
+            content: memory.to_string(),
+            recency: 1.,
+            recall: 1.,
+        };
+
+        let mut latest_point_id = self.latest_point_id.lock().await;
+        *latest_point_id += 1;
+        let point_id = latest_point_id.to_string();
+
+        let memory_json = serde_json::to_value(&memory_struct)?;
+        let embedding_json = serde_json::to_value(&embedding)?;
+
+        redis::cmd("JSON.SET")
+            .arg(&point_id)
+            .arg("$")
+            .arg(format!(r#"{{"memory": {}, "embedding": {}}}"#, memory_json, embedding_json))
+            .query_async(&mut con)
+            .await?;
+
+        Ok(())
+    }
+
+    async fn get_memory_pool(&mut self, llm: &LLM, memory: &str, min_count: usize) -> Result<Vec<RelevantMemory>, Box<dyn Error>> {
+        let embedding = llm.model.get_base_embed(memory).await?;
+        let mut con = self.client.get_tokio_connection().await?;
+
+        let query_blob: Vec<u8> = embedding
+            .iter()
+            .flat_map(|&value| value.to_le_bytes().to_vec())
+            .collect();
+
+        let result: redis::Value = redis::cmd("FT.SEARCH")
+            .arg(&self.index_name) // Replace with the actual index name
+            .arg("DIALECT")
+            .arg(2)
+            .arg(format!("($query)=>[KNN {} @v $B]", min_count))
+            .arg("PARAMS")
+            .arg("query")
+            .arg(&query_blob)
+            .query_async(&mut con)
+            .await?;
+
+        let result_pairs: Vec<(String, f32)> = match result {
+            redis::Value::Bulk(items) => {
+                items
+                    .chunks_exact(2)
+                    .filter_map(|chunk| match (chunk.get(0), chunk.get(1)) {
+                        (Some(redis::Value::Data(key)), Some(redis::Value::Data(value))) => {
+                            let score: f32 = String::from_utf8_lossy(value)
+                                .parse()
+                                .unwrap_or_default();
+                            Some((String::from_utf8_lossy(key).into_owned(), score))
+                        }
+                        _ => None,
+                    })
+                    .collect()
+            }
+            _ => vec![],
+        };
+
+        let mut relevant_memories = vec![];
+        for (key, similarity) in result_pairs {
+            let json_data: String = redis::cmd("JSON.GET")
+                .arg(&key)
+                .query_async(&mut con)
+                .await?;
+            let data: serde_json::Value = serde_json::from_value(serde_json::Value::String(json_data))?;
```

Comment:
This would be another example where use of some sort of `EmbeddedMemory` struct would help clarify the code, and also eliminate unsafe `get` operations from the `data` object.